src/query/functions/binary/node.go

```diff
@@ -0,0 +1,482 @@
+	DivType = "/"
+
+	// ExpType raises lhs to the power of rhs
+	// NB: to keep consistency with prometheus (and go)
+	//  0 ^ 0 = 1
+	//  NaN ^ 0 = 1
+	ExpType = "^"
+
+	// ModType takes the modulo of lhs by rhs
+	// Special cases are:
+	// 	 X % 0 = NaN
+	// 	 NaN % X = NaN
+	// 	 X % NaN = NaN
+	ModType = "%"
+
+	// *******************************************
+	// * Comparison functions
+
+	// EqType raises lhs to the power of rhs
+	EqType = "=="
+
+	// NotEqType raises lhs to the power of rhs
+	NotEqType = "!="
+
+	// GreaterType raises lhs to the power of rhs
+	GreaterType = ">"
+
+	// LesserType raises lhs to the power of rhs
+	LesserType = "<"
+
+	// GreaterEqType raises lhs to the power of rhs
+	GreaterEqType = ">="
+
+	// LesserEqType raises lhs to the power of rhs
+	LesserEqType = "<="
+)
+
+type mathFunc func(x, y float64) float64
+
+type singleScalarFunc func(x float64) float64
+
+func toFloat(b bool) float64 {
+	if b {
+		return 1
+	}
+	return 0
+}
+
+var (
+	mathFuncs = map[string]mathFunc{
+		PlusType:     func(x, y float64) float64 { return x + y },
+		MinusType:    func(x, y float64) float64 { return x - y },
+		MultiplyType: func(x, y float64) float64 { return x * y },
+		DivType:      func(x, y float64) float64 { return x / y },
+		ModType:      math.Mod,
+		ExpType:      math.Pow,
+
+		EqType:        func(x, y float64) float64 { return toFloat(x == y) },
+		NotEqType:     func(x, y float64) float64 { return toFloat(x != y) },
+		GreaterType:   func(x, y float64) float64 { return toFloat(x > y) },
+		LesserType:    func(x, y float64) float64 { return toFloat(x < y) },
+		GreaterEqType: func(x, y float64) float64 { return toFloat(x >= y) },
+		LesserEqType:  func(x, y float64) float64 { return toFloat(x <= y) },
+	}
+
+	errLeftScalar              = errors.New("expected left scalar but node type incorrect")
+	errRightScalar             = errors.New("expected right scalar but node type incorrect")
+	errNoModifierForComparison = errors.New("comparisons between scalars must use BOOL modifier")
+	errNoMatching              = errors.New("functions must have a vector matching set")
+)
+
+type binaryOp struct {
+	OperatorType string
+	fn           mathFunc
+	info         NodeInformation
+}
+
+// OpType for the operator
+func (o binaryOp) OpType() string {
+	return o.OperatorType
+}
+
+// String representation
+func (o binaryOp) String() string {
+	return fmt.Sprintf("type: %s, lnode: %s, rnode: %s", o.OpType(), o.info.LNode, o.info.RNode)
+}
+
+// Node creates an execution node
+func (o binaryOp) Node(controller *transform.Controller) transform.OpNode {
+	return &binaryNode{
+		controller: controller,
+		cache:      transform.NewBlockCache(),
+		op:         o,
+	}
+}
+
+type binaryNode struct {
+	op         binaryOp
+	controller *transform.Controller
+	cache      *transform.BlockCache
+	mu         sync.Mutex
+}
+
+// NodeInformation describes the types of nodes used
+// for binary operations
+type NodeInformation struct {
```

Comment:
node information is too generic :(  how about binaryOpParams or something more specific ?

Reply:
It's in the binary package, so it would be binary.NodeInformation if we're using it elsewhere? If you're still against it can change thi

Reply:
Met you halfway with NodeParams :p